File: problems/1901_find_a_peak_element_ii/Solution.java
class Solution {
    public int[] findPeakGrid(int[][] mat) {
        int m = mat.length, n = mat[0].length;
        int lo = 0, hi = m - 1;
        while (lo <= hi) {
            int mid = (lo + hi) / 2;
            int maxCol = 0;
            for (int j = 1; j < n; j++) if (mat[mid][j] > mat[mid][maxCol]) maxCol = j;
            int above = mid > 0 ? mat[mid - 1][maxCol] : -1;
            int below = mid < m - 1 ? mat[mid + 1][maxCol] : -1;
            if (mat[mid][maxCol] > above && mat[mid][maxCol] > below) return new int[]{mid, maxCol};
            else if (above > mat[mid][maxCol]) hi = mid - 1;
            else lo = mid + 1;
        }
        return new int[]{-1, -1};
    }
}

File: problems/1901_find_a_peak_element_ii/solution.py
from typing import List


class Solution:
    def findPeakGrid(self, mat: List[List[int]]) -> List[int]:
        m, n = len(mat), len(mat[0])
        lo, hi = 0, m - 1
        while lo <= hi:
            mid = (lo + hi) // 2
            max_col = max(range(n), key=lambda j: mat[mid][j])
            above = mat[mid - 1][max_col] if mid > 0 else -1
            below = mat[mid + 1][max_col] if mid < m - 1 else -1
            if mat[mid][max_col] > above and mat[mid][max_col] > below:
                return [mid, max_col]
            elif above > mat[mid][max_col]:
                hi = mid - 1
            else:
                lo = mid + 1
        return [-1, -1]
